power_grid_model_c/power_grid_model/include/power_grid_model/sparse_idx_vector.hpp

```diff
@@ -0,0 +1,112 @@
+// SPDX-FileCopyrightText: 2022 Contributors to the Power Grid Model project <dynamic.grid.calculation@alliander.com>
+//
+// SPDX-License-Identifier: MPL-2.0
+
+#pragma once
+#ifndef POWER_GRID_MODEL_SPARSE_IDX_VECTOR_HPP
+#define POWER_GRID_MODEL_SPARSE_IDX_VECTOR_HPP
+
+#include "power_grid_model.hpp"
+
+#include <boost/iterator/iterator_facade.hpp>
+#include <boost/range.hpp>
+#include <boost/range/combine.hpp>
+#include <boost/range/counting_range.hpp>
+
+/*
+A data-structure for iterating through the indptr, ie. sparse representation of data.
+Indptr can be eg: [0, 3, 6, 7]
+This means that:
+objects 0, 1, 2 are coupled to index 0
+objects 3, 4, 5 are coupled to index 1
+objects 6 is coupled to index 2
+
+Another intuitive way to look at this for python developers is like list of lists: [[0, 1, 2], [3, 4, 5], [6]].
+
+DenseIdxVector is a vector of element to group. ie. [0, 1, 1, 4] would denote that [[0], [1, 2], [], [], [3]]
+The input, ie. [0, 1, 3] should be strictly increasing
+
+*/
+
+namespace power_grid_model::detail {
+
+class SparseIdxVector {
+  public:
+    explicit SparseIdxVector(IdxVector indptr) : indptr_(indptr) {}
+
+    template <class Value>
+    class Iterator : public boost::iterator_facade<Iterator<Value>, Value, boost::random_access_traversal_tag,
+                                                   boost::iterator_range<IdxCount>, Idx> {
+      public:
+        explicit Iterator(IdxVector indptr, Idx idx) : indptr_(indptr), idx_(idx) {}
+
+      private:
+        IdxVector indptr_;
+        Idx idx_;
+        friend class boost::iterator_core_access;
+
+        boost::iterator_range<IdxCount> dereference() const {
+            return boost::counting_range(indptr_[idx_], indptr_[idx_ + 1]);
+        }
+        bool equal(Iterator const& other) const { return idx_ == other.idx_; }
+        void increment() { ++idx_; }
+        void decrement() { --idx_; }
+        void advance(Idx n) { idx_ += n; }
+        Idx distance_to(Iterator const& other) const { return other.idx_ - idx_; }
+    };
+
+    auto get_element_range(Idx group) { return boost::iterator_range<IdxCount>(indptr_[group], indptr_[group + 1]); }
+
+    Idx get_group(Idx element) {
+        return static_cast<Idx>(std::upper_bound(indptr_.begin(), indptr_.end(), element) - indptr_.begin());
```

Comment:
i think it should throw an `std::out_of_range` error when reading outside the allowed elements

https://en.cppreference.com/w/cpp/error/out_of_range